### PR TITLE
Add fairness report

### PR DESCRIPTION
**Pull request by nsorros**

### CODE REVIEW

File: skfair/metrics/fairness_report.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            'PPVR': precision,
+            'FDR': fp / (tp + fp + eps),
+            'ACC': (tp + tn) / (tp + fp + tn + fn + eps),
+            'F1': 2 * precision * recall / (precision + recall + eps),
+            'Support': len(group_data)
+        })
+
+    headers = outputs[0].keys()
+
+    # almost identical to from https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/_classification.py
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:37]: that's good :)

File: skfair/metrics/fairness_report.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    grouped_data = defaultdict(list)
+    for i, (y_t, y_p) in enumerate(zip(y_true, y_pred)):
+        group_key = group_names[i] if group_names else groups[i]
+        grouped_data[group_key].append((y_t, y_p))
+
+    outputs = []
+    for group_name, group_data in grouped_data.items():
+        y_true_group, y_pred_group = zip(*group_data)
+
+        conf_matrix = confusion_matrix(y_true_group, y_pred_group, labels=[0,1])
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:18]: It is possible that we have different labels, not just `[0, 1]`. Possibly an option to add as an optional parameter? 

Come to think of it ... you may need to call this `classification_fairness_report`. We report for `regression` can also have fairness in it.

**Review by koaning — CHANGES_REQUESTED** [review]

**Comment by koaning** [comment]
If possible, could you check what flake8 is doing locally? CI seems to spot a lot of missing whitespace. If you're unfamiliar with flake8 this resource may help; https://calmcode.io/flake8/introduction.html

```
 skfair/metrics/fairness_report.py:6:1: E302 expected 2 blank lines, found 1
skfair/metrics/fairness_report.py:7:5: F841 local variable 'unique_groups' is assigned to but never used
skfair/metrics/fairness_report.py:18:77: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:50:1: E305 expected 2 blank lines after class or function definition, found 1
skfair/metrics/fairness_report.py:51:25: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:51:27: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:51:29: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:51:31: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:52:25: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:52:27: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:52:29: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:52:31: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:53:25: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:53:27: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:53:29: E231 missing whitespace after ','
skfair/metrics/fairness_report.py:53:31: E231 missing whitespace after ','
```

File: skfair/metrics/fairness_report.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        tn, fp, fn, tp = conf_matrix.ravel()
+        eps = 1e-10
+        precision = tp / (tp + fp + eps)
+        recall = tp / (tp + fn + eps)
+
+        outputs.append({
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:24]: I'm curious if we want to allow the users to subset/add metrics that are of interest here. @MBrouns got thoughts on this? I don't want to overload the parameters but it feels like there may be some valid wiggle room here.

File: skfair/metrics/fairness_report.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from collections import defaultdict
+
+from sklearn.metrics import confusion_matrix
+import numpy as np
+
+def fairness_report(y_true, y_pred, groups, group_names=None):
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:6]: This may be a nice to have, and I'm not sure, but @MBrouns @nsorros would it make sense that we also allow the user to receive a pandas dataframe as output?

File: skfair/metrics/fairness_report.py (hunk 3 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:24]: that is a good point, any thoughts on a set of metrics that we want to allow?

File: skfair/metrics/fairness_report.py (hunk 2 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:18]: 100% we can do this, and actually this is how I have implemented it internally. in the case of non binary labels, there is a slight modification needed to get tp, fp etc from the confusion matrix. Let me add this.

File: skfair/metrics/fairness_report.py (hunk 3 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py:24]: I may need to think about this. One attitude is that this is not set in stone so this is not a blocker. But looking at the [documentation on metrics](https://scikit-fairness.netlify.app/api/metrics.html) we currently support `p_percent_score` and `equal_opportunity`. It may be that you only care about one, or both. In the future I expect this list to grow too. So it feels like we should have some flexbility but at the same time I don't want to make it easy for people to over-rule a good standard. There's also the concern that you may want to keep functions simple.

File: skfair/metrics/fairness_report.py (hunk 2 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py:18]: Awesome :)

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:6]: makes sense, actually in the original classification report they have an argument to return a dict so that or pandas makes sense. will add.

Comment (koaning) [reply skfair/metrics/fairness_report.py:6]: I've justed checked their function signature. Where possible it makes sense to copy it as much as possible so we make sure that we adhere to their standard. I also see they have `target_names=None, sample_weight=None, digits=2`. For now I think target names would be good to have. `sample_weight` and `digits` can be saved for later.

**Comment by nsorros** [comment]
> If possible, could you check what flake8 is doing locally? CI seems to spot a lot of missing whitespace. If you're unfamiliar with flake8 this resource may help; https://calmcode.io/flake8/introduction.html
> 
> ```
>  skfair/metrics/fairness_report.py:6:1: E302 expected 2 blank lines, found 1
> skfair/metrics/fairness_report.py:7:5: F841 local variable 'unique_groups' is assigned to but never used
> skfair/metrics/fairness_report.py:18:77: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:50:1: E305 expected 2 blank lines after class or function definition, found 1
> skfair/metrics/fairness_report.py:51:25: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:51:27: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:51:29: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:51:31: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:52:25: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:52:27: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:52:29: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:52:31: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:53:25: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:53:27: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:53:29: E231 missing whitespace after ','
> skfair/metrics/fairness_report.py:53:31: E231 missing whitespace after ','
> ```

I am familiar with it, will check. I thought it run automatically as a pre commit hook.

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py]: Added `output_dict` and `labels`. In regards to target_names, should we rename group_names to target_names? Target names is slighly different for this function as we do not care so much about the names of the target variables, more so about the group names or names for the sensitive characteristic. As such I am wondering whether we should differentiate the name to make the distinction clear.

File: skfair/metrics/fairness_report.py (hunk 3 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py]: I think it is a good idea to customise the metrics as the metrics you might want to track vary on a case by case basis in fairness. Should we add a parameter metrics that accepts a list with string names for those metrics we support? Or should I just add these two metrics into the report for now?

**Comment by koaning** [comment]
Feel free to add flake8 in pre-commit if it is missing.

File: skfair/metrics/fairness_report.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        tp += conf_matrix[i, i]
+        fn += conf_matrix[i, :].sum() - conf_matrix[i, i]
+        fp += conf_matrix[:, i].sum() - conf_matrix[i, i]
+        tn += (conf_matrix[:i-1, :i-1].sum() + conf_matrix[:i-1, i+1:].sum() +
+               conf_matrix[i+1:, :i-1].sum() + conf_matrix[i+1:i+1:].sum())
+    return tn, fp, fn, tp
+
+
+def classification_fairness_report(y_true, y_pred, groups, group_names=None,
+                                   labels=None, output_dict=False):
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:27]: Would it not be preferable here to have `output` instead of `output_dict`? 

This way the user can specify `output="dict"`, `output="text"` or `output="pandas"`.

File: skfair/metrics/fairness_report.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            'PPVR': precision,
+            'FDR': fp / (tp + fp + eps),
+            'ACC': (tp + tn) / (tp + fp + tn + fn + eps),
+            'F1': 2 * precision * recall / (precision + recall + eps),
+            'Support': len(group_data)
+        }
+
+    if output_dict:
+        return report_dict
+    else:
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:55]: No need for the `else` here.

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py]: It may be that we care about `target` names though. Let's say I am worried about `gender` group and I have an income prediction usecase. There may be unfairness but also inaccuracy. I may be worried about the `low_income` target and the `high_income` target but not the `medium_income` target.

File: skfair/metrics/fairness_report.py (hunk 3 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py]: Why not a list that accepts functions? That way the user can also pass in their own custom function. Or a dictionary with names and functions?

Comment (nsorros) [reply skfair/metrics/fairness_report.py]: Sounds good, will implement

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py]: Good point, will add.

**Comment by nsorros** [comment]
> Feel free to add flake8 in pre-commit if it is missing.

Will check and add if missing

**Comment by nsorros** [comment]
@koaning should i open the PR btw or leave in draft?

File: skfair/metrics/fairness_report.py (hunk 5 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:27]: output_dict is the sklearn parameter, your suggestions sounds better, let's change to that.

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py]: Actually to clarify what is the intended behaviour here? Is it to filter medium income and calculate fairness metrics on the other two? Or to display them by breaking down the sensitive variable by target name, a multi group of some sort where you see the metrics per gender per income_group?

File: skfair/metrics/fairness_report.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    for group_name, group_data in grouped_data.items():
+        y_true_group, y_pred_group = zip(*group_data)
+
+        for metric_name, metric in yield_metrics(metrics):
+            report_dict[group_name][metric_name] = metric(y_true_group, y_pred_group, labels)
+        report_dict[group_name]["Support"] = len(group_data)
+
+    if output == "dict":
+        return report_dict
+    elif output == "pandas":
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:80]: `elif` can also just be `if`

File: skfair/metrics/fairness_report.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+def true_false_positive_negative(conf_matrix):
+    # conf matrix: rows are true, cols are pred
+    if conf_matrix.shape == (2, 2):  # binary case
+        tn, fp, fn, tp = conf_matrix.ravel()
+        return tn, fp, fn, tp
+    tp = 0
+    fp = 0
+    fn = 0
+    tn = 0
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:42]: Do we *need* this code here? it seems that the code below handles the general case and I cannot imagine that there's such a substantial performance boost.

File: skfair/metrics/fairness_report.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    tn, fp, fn, tp = true_false_positive_negative(conf_matrix)
+    eps = 1e-10
+    return fp / (fp + tn + eps)
+
+
+def false_discovery_score(y_true, y_pred, labels=None):
+    conf_matrix = confusion_matrix(y_true, y_pred, labels=labels)
+    tn, fp, fn, tp = true_false_positive_negative(conf_matrix)
+    eps = 1e-10
+    return fp / (tp + fp + eps)
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:21]: These are nice to have. Preferably we'd have these metrics in a seperate file though. You could also add a unit test/docstring but that can also be left for another PR.

File: skfair/metrics/fairness_report.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    report = classification_fairness_report(y_true, y_pred, groups, group_names, output="dict")
+    print(report)
+
+    y_true = np.array([0, 0, 1, 1, 1, 2, 2])
+    y_pred = np.array([0, 0, 0, 1, 1, 2, 0])
+    groups = np.array([1, 0, 1, 0, 1, 0, 1])
+    group_names = ["b", "r", "b", "r", "b", "r", "b"]
+
+    report = classification_fairness_report(y_true, y_pred, groups, group_names, labels=[0, 1, 2])
+    print(report)
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:117]: Since the code is starting to become PR-ready. It might be good to move some of these to a test file and to add some unit tests.

File: skfair/metrics/fairness_report.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        tn += (conf_matrix[:i-1, :i-1].sum() + conf_matrix[:i-1, i+1:].sum() +
+               conf_matrix[i+1:, :i-1].sum() + conf_matrix[i+1:i+1:].sum())
+    return tn, fp, fn, tp
+
+
+def yield_metrics(metrics):
+    if type(metrics) == list:
+        for metric in metrics:
+            yield metric.__name__, metric
+    else:
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:57]: Might be nicer to check for `list` and `dict` and to throw an error otherwise. This is also a function that can be hidden away from the end-user.

File: skfair/metrics/fairness_report.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    width = max(len(cn) for cn in grouped_data.keys())
+    head_fmt = '{:>{width}s} ' + ' {:>10}' * (len(headers))
+    report = head_fmt.format('', *headers, width=width)
+    report += '\n\n'
+    row_fmt = '{:>{width}s} ' + ' {:>10.{digits}f}' * (len(headers)-1) + ' {:>10}\n'
+    for row_name, row in report_dict.items():
+        row_values = [row_name] + list(row.values())
+        report += row_fmt.format(*row_values, width=width, digits=3)
+
+    report += '\n'
```
Comment (koaning) [inline skfair/metrics/fairness_report.py:95]: Could we maybe make a function out of this? Something like `_into_terminal_table`? That should make it slightly more maintainable.

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -8,6 +8,7 @@ repos:
     -   id: check-docstring-first
     -   id: check-merge-conflict
     -   id: check-added-large-files
+    -   id: flake8
```
Comment (koaning) [inline .pre-commit-config.yaml:11]: Nice addition.

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py]: I think I had the term `label` mixed up with `target`. But there is a discussion to be had on something that is very related.

This is the current output of two calls to the summary. 

```
       RECALL        FPR       PPVR        FDR        ACC         F1    Support

b       0.667      0.000      0.667      0.000      2.000      0.667          3
r       1.000      0.000      1.000      0.000      2.000      1.000          2


       RECALL        FPR       PPVR        FDR        ACC         F1    Support

b       0.500      0.000      0.333      0.000      0.333      0.400          3
r       1.000      0.000      1.000      0.000      1.000      1.000          4
```

Is it clear what the difference is? This was the call;

```python
y_true = np.array([0, 0, 1, 1, 1, 2, 2])
y_pred = np.array([0, 0, 0, 1, 1, 2, 0])
groups = np.array([1, 0, 1, 0, 1, 0, 1])
group_names = ["b", "r", "b", "r", "r", "r", "b"]

report = classification_fairness_report(
  y_true, y_pred, 
  groups, group_names, labels=[0, 2]
)
print(report)
report = classification_fairness_report(
  y_true, y_pred, 
  groups, group_names, labels=[0, 1]
)
print(report)
```

It would be nice to incorporate the labels into the report somehow.

Comment (koaning) [reply skfair/metrics/fairness_report.py]: @MBrouns am very curious to hear your insights here. I'm wondering what table layout is going to be the "most obvious".

File: skfair/metrics/fairness_report.py (hunk 12 above)
Comment (MBrouns) [reply skfair/metrics/fairness_report.py:95]: I'm also not opposed to adding a dependency such as [terminal tables](https://pypi.org/project/terminaltables/) for these kinds of things

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (MBrouns) [reply skfair/metrics/fairness_report.py]: I think the default sklearn implementation has a row per target right? Considering we want it also per group, would a 'nested' table like a pandas multi-index  make sense?

Comment (koaning) [reply skfair/metrics/fairness_report.py]: @MBrouns This was indeed the direction that I was wondering about. Scikit-learn only concerns itself with reporting per class, since we've added a group for fairness it may also be valid to think on what the most obvious way is to represent both in this summary. 

I'm guessing adding a column/index makes sense. Since the group is not manditory it may be valid to consider it secondary? So first index the label value and the second one the group? Then again ... if the group is given it may be cosidered more important? It's not 100% clear to me what the best behavior is here.

Comment (MBrouns) [reply skfair/metrics/fairness_report.py]: I feel like in general i'd be more interested in comparing different groups for a single outcome instead of comparing the same group for different outcomes. That'd mean that putting the label first would make more sense

File: skfair/metrics/fairness_report.py (hunk 9 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:21]: 👍

File: skfair/metrics/fairness_report.py (hunk 10 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:117]: Definitely 👍

File: skfair/metrics/fairness_report.py (hunk 11 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:57]: You mean by adding an underscore in the start of the name?

File: skfair/metrics/fairness_report.py (hunk 12 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:95]: Happy with both, will start by moving into a separate function but happy to move to terminal tables if there is consensus

File: skfair/metrics/fairness_report.py (hunk 8 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:42]: You mean removing these lines right?
```
if conf_matrix.shape == (2, 2):  # binary case
        tn, fp, fn, tp = conf_matrix.ravel()
        return tn, fp, fn, tp
```

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py]: I agree with @MBrouns, I would also be more keen to compare groups easily thus putting label first.

Again to clarify, what is the intended behaviour when labels=None? Do we print the labels and groups or compute the metrics across all labels?

File: skfair/metrics/fairness_report.py (hunk 8 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py:42]: Ah sorry, I selected too many lines there. But yes :)

File: skfair/metrics/fairness_report.py (hunk 12 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py:95]: Something tells me that terminal tables are going to be easier to maintain.

File: skfair/metrics/fairness_report.py (hunk 4 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py]: It feels to me that being as explicit as possible is the safest here. We wouldn't want the wrong decision being made and by showing both the label and the group the user can check if they understand the table. By computing the metric across all labels, a user might make a wrong assumption.

File: skfair/metrics/fairness_report.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        for metric_name, metric in _yield_metrics(metrics):
+            report_dict[group_name][metric_name] = metric(y_true_group, y_pred_group, labels)
+        report_dict[group_name]["Support"] = len(group_data)
+
+    if output == "dict":
+        return report_dict
+    if output == "pandas":
+        return pd.DataFrame(report_dict)
+
+    headers = [v.keys() for k, v in report_dict.items()][0]
```
Comment (nsorros) [inline skfair/metrics/fairness_report.py:56]: should we still move this to a separate function?

File: skfair/metrics/fairness_report.py (hunk 10 above)
Comment (nsorros) [reply skfair/metrics/fairness_report.py:82]: Aim to write tests tomorrow morning and I guess then we can open this PR for further review?

File: skfair/metrics/fairness_report.py (hunk 14 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py:56]: That would be better. Feel free to keep it in the same file though.

File: skfair/metrics/fairness_report.py (hunk 10 above)
Comment (koaning) [reply skfair/metrics/fairness_report.py:82]: Done.

**Comment by nsorros** [comment]
Wrote some tests, happy to write some more if needed for some of the use cases.

Notice I reintroduced the 2x2 special case for the confusion matrix, this is because in the 2x2 case we do not treat each row or column as separate class which makes the computation different i.e. we do not sum the diagonal to get true positives etc.

**Review by koaning — APPROVED** [review]

**Comment by koaning** [comment]
This looks good to me. Thanks! 

I'll merge this now but I should point out that we've been made aware of another project with similar goals. See [here](https://github.com/fairlearn/fairlearn/issues/406). We're currently discussing with them what the best way forward is. Collaboration might make sense, two packages might also make sense.

I'll keep you in the loop but thanks for this!

**Comment by nsorros** [comment]
> This looks good to me. Thanks!
> 
> I'll merge this now but I should point out that we've been made aware of another project with similar goals. See [here](https://github.com/fairlearn/fairlearn/issues/406). We're currently discussing with them what the best way forward is. Collaboration might make sense, two packages might also make sense.
> 
> I'll keep you in the loop but thanks for this!

Awesome, keep me in the loop indeed. I would love to keep contributing and helping with the package.
